graph.go

```diff
@@ -7,156 +7,31 @@ type Graph[K comparable, T any] interface {
-	BFS(start T, visit func(value T) bool) error
-
-	// BFSByHash does the same as BFS, but uses a hash value to identify the starting vertex.
-	BFSByHash(startHash K, visit func(value T) bool) error
-
-	// CreatesCycle determines whether an edge between the given source and target vertices would
-	// introduce a cycle. It won't create that edge in any case.
-	//
-	// A potential edge would create a cycle if the target vertex is also a parent of the source
-	// vertex. Given a graoh A-B-C-D, adding an edge DA would introduce a cycle:
-	//
-	//	A -
-	//	|  |
-	//	B  |
-	//	|  |
-	//	C  |
-	//	|  |
-	//	D -
+	// AddEdge creates an edge between the source and the target vertex. If the Directed option has
+	// been called on the graph, this is a directed edge. Returns an error if either vertex doesn't
+	// exist or the edge already exists.
 	//
-	// CreatesCycle backtracks the ingoing edges of D, resulting in a reverse walk C-B-A.
-	CreatesCycle(source, target T) (bool, error)
-
-	// CreatesCycleByHashes does the same as CreatesCycle, but uses a hash value to identify the
-	// starting vertex.
-	CreatesCycleByHashes(sourceHash, targetHash K) (bool, error)
-
-	// Degree determines and returns the degree of a given vertex.
-	Degree(vertex T) (int, error)
-
-	// DegreeByHash does the same as Degree, but uses a hash value to identify the vertex.
-	DegreeByHash(vertexHash K) (int, error)
-
-	// StronglyConnectedComponents detects all strongly connected components within the graph and
-	// returns the hash values of the respective vertices for each component. This only works for
-	// directed graphs.
+	// AddEdge accepts a variety of functional options to set further edge details such as the
+	// weight or an attribute:
 	//
-	// Note that the current implementation uses recursive function calls.
-	StronglyConnectedComponents() ([][]K, error)
-
-	// ShortestPath computes the shortest path between a source and a target vertex using the edge
-	// weights and returns the hash values of the vertices forming that path. This search runs in
-	// O(|V|+|E|log(|V|)) time.
+	//	_ = graph.Edge("A", "B", graph.EdgeWeight(4), graph.EdgeAttribute("label", "my-label"))
 	//
-	// The returned path includes the source and target vertices. If the target cannot be reached
-	// from the source vertex, ShortestPath returns an error. If there are multiple shortest paths,
-	// an arbitrary one will be returned.
-	ShortestPath(source, target T) ([]K, error)
+	AddEdge(sourceHash, targetHash K, options ...func(*EdgeProperties)) error
 
-	// ShortestPathByHashes does the same as ShortestPath, but uses hash values to identify the
-	// vertices.
-	ShortestPathByHashes(sourceHash, targetHash K) ([]K, error)
+	// Edge returns the edge between two vertices. The second return value indicates whether the
+	// edge exists. If the graph is undirected, an edge with swapped source and target vertices
+	// does match.
+	Edge(sourceHash, targetHash K) (Edge[T], bool)
```

Comment:
`Edge()` seems to be the only method that deals with `Edge[T]`, would it be a bit more consistent to return `Edge[K]` as `AdjacencyMap()` and `Predecessors()` do in some way?

In the current codebase the result of `Edge()` isn't really used anywhere.

Reply:
This is a tradeoff between efficiency and developer-friendliness in my opinion. The initial reason I made this an `Edge[T]` is that it is convenient to work with the retrieved vertex values directly. If we returned an `Edge[K]`, the vertex value would need to be looked up first. Tbh, I'm not sure which approach is better here. Maybe another function `EdgeWithValues(source, target K) (Egde[T], error)` would make sense, but that would be yet another method...

Reply:
I think adding `EdgeWithValues(source, target K) (Egde[T], error)` to imprve the developer experiece while also providing `Edge(source, target K) (Egde[K], error)` for performance would be really appreciated imo.

Reply:
Owh one more thing before I forget about it.

`Edge(sourceHash, targetHash K) (Edge[T], bool)` (and `EdgeWithValues()`) should probably also return an error, in cases developers need to differentiate between "not found" and "can't connect to storage".

Additionally for `EdgeWithValues()`, one of the two vertices might not exist in the store, even though the edge might exist, so an error should be provided.

I'm not sure if "not found" should be returned as a bool still, or as a specific error (ie `ErrEdgeNotFound`, `ErrVertexNotFound`).

Reply:
Yes, this is exactly what I'm currently working on. I prefer creating a predefined `ErrEdgeNotFound` instance over an additional boolean value, because handling 3 return values might get cumbersome - especially if you can safely assume that the edge exists.

```go
// Having to ignore two return values is a bit tedious.
edge, _, _ := g.Edge("munich", "london")
 ```

Also, returning a predefined error instance is consistent with the standard library, e.g. `sql.ErrNoRows`.

Reply:
I changed `Edge` to return an error (and `ErrEdgeNotFound` in particular, if the edge couldn't be found).